fix(dev): add `hasLatestBuildOutput` instead of `scheduleBuildIfStale` and add edit-reload test

sapphi-red opened the pull request:
The following scenario was not working.

1. Edit file
2. HMR happens
3. Reload the window (stale content was shown!)

It is because the fix for `ensure_latest_build_output` in #6262 wasn't applied to `scheduleBuildIfStale`. To achieve that properly, it requires a significant refactor as `scheduleBuildIfStale` returns a future of the build process. For now, I added `hasLatestBuildOutput` instead to achieve a similar behavior. This approach has a problem of race condition because the internal state mutex is released in between the `hasLatestBuildOutput` call and the `ensureLatestBuildOutput` call. But this should be good enough for the initial implementation.

sapphi-red commented:
* **#6321** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6321?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6321?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>graphite: merge</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/rolldown?ref=merge-queue-instructions-in-stack-comment&prId=6321).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

hyf0 commented:
Good one. I plan to finish https://github.com/rolldown/rolldown/pull/6297 today. It's a big change. I want to delay this before we merge it, so we could see if there's any conflict we need to solve. I could solve conflicts, just want to check how they interacts.

hyf0 commented:
### Merge activity

* **Sep 24, 2:54 PM UTC**: The merge label 'graphite: merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown) once it meets the requirements.
* **Sep 24, 2:56 PM UTC**: `hyf0` added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown).
* **Sep 24, 3:09 PM UTC**: Merged by the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown).

sapphi-red committed:
fix(dev): add `hasLatestBuildOutput` instead of `scheduleBuildIfStale` and add edit-reload test (#6321)

The following scenario was not working.

1. Edit file
2. HMR happens
3. Reload the window (stale content was shown!)

It is because the fix for `ensure_latest_build_output` in #6262 wasn't applied to `scheduleBuildIfStale`. To achieve that properly, it requires a significant refactor as `scheduleBuildIfStale` returns a future of the build process. For now, I added `hasLatestBuildOutput` instead to achieve a similar behavior. This approach has a problem of race condition because the internal state mutex is released in between the `hasLatestBuildOutput` call and the `ensureLatestBuildOutput` call. But this should be good enough for the initial implementation.
